### ISSUE TITLE
Framework: App Shell

### ISSUE THREAD

**Issue by apeatling** [Automattic/wp-calypso#19833]
Rendering the shell of the app on the server helps with the perception of speed.

Intricacies include not knowing specifics of the layout for a given section — sidebar, full-width, etc. We are already rendering an empty layout which is good. Extending this to include user-based information (like the masterbar) requires solving hurdles around internationalization, like having to avoid displaying text until we can use i18n-calypso on the server side. 

The other aspect is looking at [shipping critical CSS](https://github.com/Automattic/wp-calypso/pull/15464) for the shell earlier.

**Comment by flootr** [comment]
After some thoughts I think there are some steps needed before this issue actually can be resolved:

- [ ] #19822 **Eliminate Jade**: I think this should be done before to save work. We would have to refactor the work on the app shells if we change the way anything got tied together on the server afterwards. @ockham [worked on this before](https://github.com/Automattic/wp-calypso/pull/8459), would it be okay for you if I continue the work you've begun?
- [ ] **Inline critical css**: This is actually part of this issue, but also good to be done before. If we can inline critical css and defer loading the rest of the css we can improve performance significantly. @alexsanford has [already done some work on this](https://github.com/Automattic/wp-calypso/pull/15464) which I also'd like to continue.

### Lighthouse

According to the performance audit there's currently have ~10 seconds until the first meaningful paint:

<img width="691" alt="lighthouse - performance metrics" src="https://user-images.githubusercontent.com/9202899/34600208-a5d8ce18-f1f6-11e7-9e9d-06fd82ebb3e1.png">

These stats were relatively stable during my tests. And it's mainly due to CSS blocking the rendering path:

<img width="689" alt="lighthouse - render blocking css" src="https://user-images.githubusercontent.com/9202899/34600245-c9e56b5e-f1f6-11e7-9252-d420c150235c.png">

#### Noticons

About the `noticons.css` (blocking for ~1.2s) I found this:

<img width="978" alt="don't use noticons" src="https://user-images.githubusercontent.com/9202899/34600942-27d659fa-f1fa-11e7-8425-51cb97e66496.png">

I haven't found an open issue regarding this statement, may be worth revisiting.

/cc @ehg

**Comment by stephanethomas** [comment]
Are you going to consider pages that are rendered server-side as well, such as the [`Login` page](https://wordpress.com/log-in)?

**Comment by flootr** [comment]
Hey @stephanethomas, thanks for your question. I think server side rendered pages are not in the scope of this issue, because there won't be an app shell as they're rendered completely on the server. Or is there another aspect we might need to consider?

The Login page isn't server rendered for me. Is there anything I'm missing? I get an empty layout when disabling JavaScript in the browser.

**Comment by stephanethomas** [comment]
> I think server side rendered pages are not in the scope of this issue, because there won't be an app shell as they're rendered completely on the server. Or is there another aspect we might need to consider?

Sorry, I wasn't clear at all - I should have taken more coffee. My concern was more about keeping these specific pages in mind to avoid any bad surprise since you're going to touch Jade templates or `server/pages/index.js`.

> The Login page isn't server rendered for me. Is there anything I'm missing? I get an empty layout when disabling JavaScript in the browser.

We don't render pages server-side when the user is logged in. Moreover in the case of the `Login` page we only render the default English variant that lives at https://wordpress.com/log-in. Variants for other languages (e.g. https://wordpress.com/log-in/fr) are only rendered client-side.

**Comment by flootr** [comment]
I agree this needs to be considered and will make sure to not break any server rendered content, thanks for clarification and pointing that out 👍 ☕️
